moxin-backend/src/store/remote.rs

```diff
@@ -0,0 +1,278 @@
+    let response = client.head(url).send()?;
+
+    let content_length = response
+        .headers()
+        .get(reqwest::header::CONTENT_LENGTH)
+        .and_then(|val| val.to_str().ok())
+        .and_then(|val| val.parse::<u64>().ok())
+        .unwrap_or(0);
+
+    Ok(content_length)
+}
+
+fn download_file(
+    client: &reqwest::blocking::Client,
+    content_length: u64,
+    url: &str,
+    local_path: &str,
+    step: f64,
+    report_fn: &mut dyn FnMut(f64),
+) -> io::Result<f64> {
+    use std::path::Path;
+    let path: &Path = local_path.as_ref();
+    std::fs::create_dir_all(path.parent().unwrap())?;
+    let mut file = File::options()
+        .write(true)
+        .create(true)
+        .open(local_path)
+        .unwrap();
+    let file_length = file.metadata()?.len();
+
+    if file_length < content_length {
+        file.seek(io::SeekFrom::End(0))?;
+
+        let range = format!("bytes={}-", file_length);
+        let mut resp = client
+            .get(url)
+            .header("Range", range)
+            .send()
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
+
+        let mut buffer = vec![0; (content_length as usize) / 100];
+        let mut downloaded: u64 = file_length;
+        let mut last_progress = 0.0;
+        loop {
+            let len = match resp.read(&mut buffer) {
+                Ok(0) => break, // EOF
+                Ok(len) => len,
+                Err(e) => return Err(e),
+            };
+            file.write_all(&buffer[..len])?;
+            downloaded += len as u64;
+
+            let progress = (downloaded as f64 / content_length as f64) * 100.0;
+            if progress > last_progress + step {
+                last_progress = progress;
+                report_fn(progress)
+            }
+        }
+        Ok((downloaded as f64 / content_length as f64) * 100.0)
+    } else {
+        Ok(100.0)
+    }
+}
+
+pub fn download_file_from_remote(
+    model_id: &str,
+    file: &str,
+    local_path: &str,
+    step: f64,
+    report_fn: &mut dyn FnMut(f64),
+) -> io::Result<f64> {
+    let url = format!(
+        "https://huggingface.co/{}/resolve/main/{}?download=true",
+        model_id, file
+    );
+
+    let client = reqwest::blocking::Client::new();
+    let content_length = get_file_content_length(&client, &url)
+        .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
+
+    download_file(&client, content_length, &url, local_path, step, report_fn)
+}
+
+pub fn download_file_loop(
+    sql_conn: Arc<Mutex<rusqlite::Connection>>,
+    rx: Receiver<(
+        super::models::Model,
+        super::download_files::DownloadedFile,
+        Sender<anyhow::Result<FileDownloadResponse>>,
+    )>,
+) {
+    while let Ok((model, mut file, tx)) = rx.recv() {
+        let file_id = file.id.clone();
+        let mut send_progress = |progress| {
+            let _ = tx.send(Ok(FileDownloadResponse::Progress(
+                file_id.clone(),
```

Comment:
There is a simple way to avoid passing the whole `Arc` here.

`(*file_id).clone()`

I wouldn't worry to much with having extra strings cloned for a while, because it is not much data involved.

If we do this, there is no need to add `Arc` into the protocol itself. I think we should leave frontend and backend decoupled in terms of concurrency aspects.

Reply:
I just felt that there were a lot of clones here, so I added an `Arc`.